File: sort/quicksort/algorithms/quicksort.java
public class SortRecursiveInPlace<T extends Comparable<? super T>> {
  public static void main(String[] args){
    String[] arrayOfString = {"example", "array"};
    SortRecursiveInPlace<String> stringSorter = new SortRecursiveInPlace<>();
    stringSorter.sortRecursiveWithSwap(arrayOfStrings, 0, arrayOfStrings.length - 1);
  }

  // basically standard in place approach
  // generic comes from T extending Comparable with associated compareTo function
  public void sortRecursiveWithSwap(T[] xs, int start, int end){
    if(start < end){
      int pivot = partition(xs, start, end);
      sortRecursiveWithSwap(xs, start, pivot);
      sortRecursiveWithSwap(xs, pivot + 1, end);
    }
  }

  // uniquely we swap in the partition function 
  private int partition(T[] xs, int start, int end){
    T pivot = xs[(start + end) / 2];
    int i = start--;
    int j = end++;

    while(true){
      do i++;
      while (xs[i].compareTo(pivot) < 0);

      do j--;
      while (xs[j].compareTo(pivot) > 0);

      if(i >= j) return j;

      swap(i, j)
    }
  }

  private void swap(T[] xs, int i, int j){
    T t = xs[i];
    xs[i] = xs[j];
    xs[j] = t;
  }
}

File: data_structures/sort/quick_sort/quicksort.go
package divide_and_conquer

import (
	"math/rand"
	"sort"
)

// standard partition in place swap
// in order to use generic comparison requires a comparison function Less
// which implements the sort.Interface
func partition(s sort.Interface, start int, end int, pivotIndex int) int {
	s.Swap(start, pivotIndex) // move it to the beginning
	i := start + 1
	j := end
	for i <= j {
		for i <= end && s.Less(i, start) { // pivot is now at start
			i++
		}
		for j >= start && s.Less(start, j) { // pivot is now at start
			j--
		}
		if i <= j {
			s.Swap(i, j)
			i++
			j--
		}
	}
	s.Swap(start, j)
	return j
}

func sortRecursiveWithSwap(s sort.Interface, start int, end int) {
	if start >= end {
		return
	}
	pivotIndex := partition(s, start, end, rand.Intn(end-start+1)+start)
	// we use a recursive closure rather than passing slices
	// this is because although slices are cheap (they do not copy the underlying array)
	// there is some overhead in constructing the slice object
	sortRecursiveWithSwap(s, start, pivotIndex+1)
	sortRecursiveWithSwap(s, pivotIndex+1, end)
}

func sortRecursiveInPlace(s sort.Interface) {
	sortRecursiveWithSwap(s, 0, s.Len()-1)
}
